File: src/interfaces/jobs-options.ts
import { RepeatOptions } from './repeat-options';
import { BackoffOptions } from './backoff-options';
import { ParentOptions } from './parent-options';

export interface JobsOptions {
  /**
   * Defaults to `Date.now()`
   */
  timestamp?: number;

  /**
   * Ranges from 1 (highest priority) to MAX_INT (lowest priority). Note that
   * using priorities has a slight impact on performance,
   * so do not use it if not required.
   */
  priority?: number;

  /**
   * An amount of miliseconds to wait until this job can be processed.
   * Note that for accurate delays, worker and producers
   * should have their clocks synchronized.
   */
  delay?: number;

  /**
   * The total number of attempts to try the job until it completes.
   */
  attempts?: number;

  /**
   * Repeat this job, for example based on a `cron` schedule.
   */
  repeat?: RepeatOptions;

  /**
   * Rate limiter key to use if rate limiter enabled.
   *
   * @see {@link https://docs.bullmq.io/guide/rate-limiting}
   */
  rateLimiterKey?: string;

  /**
   * Backoff setting for automatic retries if the job fails
   */
  backoff?: number | BackoffOptions;

  /**
   * If true, adds the job to the right of the queue instead of the left (default false)
   *
   * @see {@link https://docs.bullmq.io/guide/jobs/lifo}
   */
  lifo?: boolean;

  /**
   * The number of milliseconds after which the job should be
   * fail with a timeout error.
   */
  timeout?: number;

  /**
   * Parent setting for tracking parent job reference.
   */
  parent?: string | ParentOptions;

  /**
   * Override the job ID - by default, the job ID is a unique
   * integer, but you can use this setting to override it.
   * If you use this option, it is up to you to ensure the
   * jobId is unique. If you attempt to add a job with an id that
   * already exists, it will not be added.
   */
  jobId?: string;

  /**
   * If true, removes the job when it successfully completes
   * When given an number, it specifies the maximum amount of
   * jobs to keep.
   * Default behavior is to keep the job in the completed set.
   */
  removeOnComplete?: boolean | number;

  /**
   * If true, removes the job when it fails after all attempts.
   * When given an number, it specifies the maximum amount of
   * jobs to keep.
   */
  removeOnFail?: boolean | number;

  /**
   * Limits the amount of stack trace lines that will be recorded in the stacktrace.
   */
  stackTraceLimit?: number;

  /**
   * Internal property used by repeatable jobs.
   */
  prevMillis?: number;
}

File: java/src/main/java/io/bullmq/core/options/JobsOptions.java
package io.bullmq.core.options;

/**
 * Options that control how a single job is added and processed.
 */
public class JobsOptions {
    /**
     * Optional explicit job id. Cannot be "0" or start with "0:".
     */
    private String jobId;

    /**
     * Delay in milliseconds before the job becomes available.
     */
    private Long delay;

    /**
     * Job priority (1 = highest). 0 (or null) means no priority.
     */
    private Integer priority;

    /**
     * Number of total attempts to run the job until it completes.
     */
    private Integer attempts;

    /**
     * Backoff setting for automatic retries.
     */
    private BackoffOptions backoff;

    /**
     * When true the job is added to the right of the wait list (LIFO).
     */
    private Boolean lifo;

    /**
     * Creation timestamp (ms since epoch). Defaults to now.
     */
    private Long timestamp;

    /**
     * Remove the job when it completes: true to remove, a number to keep
     * that many, or a KeepJobs policy.
     */
    private Object removeOnComplete;

    /**
     * Remove the job when it fails. Same shape as removeOnComplete.
     */
    private Object removeOnFail;

    /**
     * Maximum number of log entries to keep for the job (0 = unlimited).
     */
    private Integer keepLogs;

    public String getJobId() {
        return jobId;
    }

    public void setJobId(String jobId) {
        this.jobId = jobId;
    }

    public Long getDelay() {
        return delay;
    }

    public void setDelay(Long delay) {
        this.delay = delay;
    }

    public Integer getPriority() {
        return priority;
    }

    public void setPriority(Integer priority) {
        this.priority = priority;
    }

    public Integer getAttempts() {
        return attempts;
    }

    public void setAttempts(Integer attempts) {
        this.attempts = attempts;
    }

    public BackoffOptions getBackoff() {
        return backoff;
    }

    public void setBackoff(BackoffOptions backoff) {
        this.backoff = backoff;
    }

    public Boolean getLifo() {
        return lifo;
    }

    public void setLifo(Boolean lifo) {
        this.lifo = lifo;
    }

    public Long getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(Long timestamp) {
        this.timestamp = timestamp;
    }

    public Object getRemoveOnComplete() {
        return removeOnComplete;
    }

    public void setRemoveOnComplete(Object removeOnComplete) {
        this.removeOnComplete = removeOnComplete;
    }

    public Object getRemoveOnFail() {
        return removeOnFail;
    }

    public void setRemoveOnFail(Object removeOnFail) {
        this.removeOnFail = removeOnFail;
    }

    public Integer getKeepLogs() {
        return keepLogs;
    }

    public void setKeepLogs(Integer keepLogs) {
        this.keepLogs = keepLogs;
    }

    /**
     * Creates a shallow copy of these options.
     */
    public JobsOptions clone() {
        JobsOptions clone = new JobsOptions();
        clone.jobId = this.jobId;
        clone.delay = this.delay;
        clone.priority = this.priority;
        clone.attempts = this.attempts;
        clone.backoff = this.backoff != null ? this.backoff.clone() : null;
        clone.lifo = this.lifo;
        clone.timestamp = this.timestamp;
        clone.removeOnComplete = this.removeOnComplete;
        clone.removeOnFail = this.removeOnFail;
        clone.keepLogs = this.keepLogs;
        return clone;
    }
}